File: internal/plant/reminder.go
package plant

import (
	"slices"
	"time"

	"github.com/google/uuid"
)

// Reminder covers the jobs no sensor checks: nothing confirms misting the way
// a probe confirms watering. AtHours is separate from EveryDays because
// misting twice a day cannot be said as an interval in days.
type Reminder struct {
	ID      uuid.UUID       `json:"id"`
	PlantID uuid.UUID       `json:"plant_id"`
	Kind    ObservationKind `json:"kind"`

	EveryDays int    `json:"every_days"`
	AtHours   []int  `json:"at_hours"`
	Active    bool   `json:"active"`
	Note      string `json:"note,omitempty"`

	LastSentAt *time.Time `json:"last_sent_at,omitempty"`
	CreatedAt  time.Time  `json:"created_at"`
}

// MaxReminderDays is a year. Past that it is not a reminder, it is a diary.
const MaxReminderDays = 365

// Remindable are the kinds worth nagging about: things a person does on a
// rhythm. There is no reminder to notice a symptom or to have a plant die.
var Remindable = []ObservationKind{
	ObservedWatered, ObservedMisted, ObservedFertilized, ObservedPruned, ObservedRepotted,
}

// Normalise sorts and dedupes the hours and fills the defaults, so two callers
// describing the same schedule store the same row.
func (r *Reminder) Normalise() {
	if r.EveryDays == 0 {
		r.EveryDays = 1
	}
	if len(r.AtHours) == 0 {
		r.AtHours = []int{DefaultReminderHour}
	}
	slices.Sort(r.AtHours)
	r.AtHours = slices.Compact(r.AtHours)
}

// DefaultReminderHour matches the daily digest, so one morning notification
// carries everything unless somebody asked for another time.
const DefaultReminderHour = 8

// Valid reports whether this reminder can be stored.
func (r Reminder) Valid() error {
	if !slices.Contains(Remindable, r.Kind) {
		return invalid("nothing is reminded about %q", r.Kind)
	}
	if r.EveryDays < 1 || r.EveryDays > MaxReminderDays {
		return invalid("every_days is %d, want 1 to %d", r.EveryDays, MaxReminderDays)
	}
	if len(r.AtHours) == 0 {
		return invalid("a reminder needs at least one hour to fire at")
	}
	for _, hour := range r.AtHours {
		if hour < 0 || hour > 23 {
			return invalid("at_hours has %d, want 0 to 23", hour)
		}
	}
	return nil
}

// slotSearchDays bounds the walk back through the calendar. Two days is enough
// to find the previous slot however the hours are arranged.
const slotSearchDays = 2

// LastSlot is the most recent moment this was scheduled to happen. Past a
// day's cadence the slot must clear when the thing was last done; within one
// day the hours are the schedule, so twice-daily misting fires twice.
func (r Reminder) LastSlot(lastDone *time.Time, now time.Time) (time.Time, bool) {
	// Counted in whole days from the day it was done, not in hours from the
	// minute. Watering at 08:30 must not push an 08:00 slot to the next day,
	// which is how a weekly chore quietly becomes an eight-day one.
	var earliest time.Time
	if lastDone != nil && r.EveryDays > 1 {
		earliest = midnight(lastDone.In(now.Location())).AddDate(0, 0, r.EveryDays)
	}

	hours := slices.Clone(r.AtHours)
	slices.Sort(hours)

	for back := 0; back <= slotSearchDays; back++ {
		day := now.AddDate(0, 0, -back)
		for _, hour := range slices.Backward(hours) {
			slot := time.Date(day.Year(), day.Month(), day.Day(), hour, 0, 0, 0, now.Location())
			if slot.After(now) {
				continue
			}
			if slot.Before(earliest) {
				return time.Time{}, false
			}
			return slot, true
		}
	}
	return time.Time{}, false
}

func midnight(t time.Time) time.Time {
	return time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, t.Location())
}

// Due reports whether this reminder is owed. Never done counts as due, and the
// clock runs from the last real action rather than the last notification, so
// one nobody acted on stays due instead of quietly rescheduling itself.
func (r Reminder) Due(lastDone *time.Time, now time.Time) bool {
	if !r.Active {
		return false
	}
	slot, ok := r.LastSlot(lastDone, now)
	if !ok {
		return false
	}
	return lastDone == nil || lastDone.Before(slot)
}

// AlreadySent reports whether the notification for the current slot has gone
// out, which is what stops an hourly job sending the same thing all day. It
// tracks the slot rather than the date so the evening misting still sends.
func (r Reminder) AlreadySent(lastDone *time.Time, now time.Time) bool {
	if r.LastSentAt == nil {
		return false
	}
	slot, ok := r.LastSlot(lastDone, now)
	if !ok {
		return true
	}
	return !r.LastSentAt.Before(slot)
}

File: ios/Planty/Models/Reminder.swift
import Foundation

/// A standing intent to do something to a plant on a rhythm, for the jobs no
/// sensor confirms. Mirrors internal/plant.Reminder.
struct Reminder: Codable, Sendable, Hashable, Identifiable {
    let id: UUID
    let plantID: UUID
    var kind: ObservationKind

    /// How often a day qualifies, and when on that day. Separate because a
    /// mushroom kit misted at 8 and again at 20 cannot be said in days alone.
    var everyDays: Int
    var atHours: [Int]

    var active: Bool
    var note: String?

    /// Filled in by the list, not by the record itself: whether this is owed
    /// right now is a question about observations, answered by the service.
    var lastDone: Date?
    var due: Bool?

    enum CodingKeys: String, CodingKey {
        case id
        case plantID = "plant_id"
        case kind
        case everyDays = "every_days"
        case atHours = "at_hours"
        case active
        case note
        case lastDone = "last_done"
        case due
    }
}

/// One owed occurrence of a recurring reminder. The scheduled time belongs in
/// the identity because morning and evening misting are two different jobs.
struct DueReminder: Codable, Sendable, Hashable, Identifiable {
    let reminder: Reminder
    let plant: Plant
    let lastDone: Date?
    let dueAt: Date

    var id: String { occurrenceID }
    var occurrenceID: String {
        "\(reminder.id.uuidString)|\(dueAt.timeIntervalSince1970)"
    }

    enum CodingKeys: String, CodingKey {
        case reminder
        case plant
        case lastDone = "last_done"
        case dueAt = "due_at"
    }
}

struct ReminderListResponse: Decodable, Sendable {
    let reminders: [Reminder]
    var count: Int?
}

/// What the app sends to set one. Omitting the schedule means daily at the
/// digest hour, which is what "remind me to water this" usually means.
struct NewReminder: Encodable, Sendable {
    var kind: ObservationKind
    var everyDays: Int
    var atHours: [Int]
    var active: Bool
    var note: String?

    enum CodingKeys: String, CodingKey {
        case kind
        case everyDays = "every_days"
        case atHours = "at_hours"
        case active
        case note
    }
}

extension Reminder {
    /// The kinds worth nagging about: things a person does on a rhythm. There
    /// is no reminder to notice a symptom or to have a plant die.
    static let remindable: [ObservationKind] = [
        .watered, .misted, .fertilized, .pruned, .repotted
    ]

    var schedule: String {
        let times = atHours.sorted().map(Self.hourLabel).joined(separator: " and ")
        if everyDays == 1 { return "Every day at \(times)" }
        return "Every \(everyDays) days at \(times)"
    }

    static func hourLabel(_ hour: Int) -> String {
        var components = DateComponents()
        components.hour = hour
        components.minute = 0
        let calendar = Calendar.current
        guard let date = calendar.date(from: components) else { return "\(hour):00" }

        let formatter = DateFormatter()
        formatter.locale = .current
        formatter.setLocalizedDateFormatFromTemplate("j")
        return formatter.string(from: date)
    }
}

extension DueReminder {
    var completionLabel: String { reminder.kind.completionLabel }

    var dueLine: String {
        "Due at \(dueAt.formatted(date: .omitted, time: .shortened))"
    }
}

extension ObservationKind {
    /// What to do, not what was recorded: a reminder is an instruction, and
    /// "watered" is not one.
    var instruction: String {
        switch self {
        case .watered: "Water it"
        case .misted: "Mist it"
        case .fertilized: "Feed it"
        case .pruned: "Prune it"
        case .repotted: "Repot it"
        default: rawValue.capitalized
        }
    }

    /// The confirmation button says exactly what will be written to history.
    var completionLabel: String {
        switch self {
        case .watered: "I watered it"
        case .misted: "I misted it"
        case .fertilized: "I fed it"
        case .pruned: "I pruned it"
        case .repotted: "I repotted it"
        default: "I did it"
        }
    }
}
